### PR TITLE
Log when respecting the Do-Not-Track header

### PR DESCRIPTION
**Pull request by ExplodingCabbage**
**Requirements**

- [x] I have added test coverage for new or changed functionality
- [x] I have followed the repository's [pull request submission guidelines](../blob/master/CONTRIBUTING.md#submitting-pull-requests)
- [ ] <s>I have validated my changes against all supported platform versions</s> - nope, sorry. See below.

**Related issues**

I haven't opened any.

**Motivation**

Currently, if a developer has Do-Not-Track enabled on their browser, the js-client-sdk simply silently fails to send any events to LaunchDarkly with no explanation of why. It *is* mentioned in the docs that LaunchDarkly respects Do-Not-Track, but a dev who hasn't seen that line in the docs, doesn't remember it, or incorrectly believes that their browser doesn't have Do-Not-Track enabled can still be caught out. Both I and another dev I work with burned a few hours on this.

**Describe the solution you've provided**

I've made it so that the first time an event isn't sent due to Do-Not-Track being enabled, a warning about this gets logged. That should tip off devs with Do-Not-Track enabled to the reason that stuff isn't working for them.

**Describe alternatives you've considered**

I can't think of any.

**Additional context**

I failed to test this manually. I tried to use my fork in a repo from work where we use LaunchDarkly by using `"launchdarkly-js-sdk-common": "git+https://github.com/ExplodingCabbage/js-sdk-common.git"` in my `package.json`, but the install failed with some sort of build error from js-sdk-common about a failed file rename:

```
$ yarn
➤ YN0000: ┌ Project validation
➤ YN0057: │ @curative/scheduler_v2_web: Resolutions field will be ignored
➤ YN0000: └ Completed
➤ YN0000: ┌ Resolution step
➤ YN0013: │ launchdarkly-js-sdk-common@https://github.com/ExplodingCabbage/js-sdk-common.git#commit=eb708e021fe33b38e7c4f5a4b970b7dfe718b24f can't be found in the cache and will be fetched from GitHub
➤ YN0013: │ launchdarkly-js-sdk-common@https://github.com/ExplodingCabbage/js-sdk-common.git#commit=eb708e021fe33b38e7c4f5a4b970b7dfe718b24f can't be found in the cache and will be fetched from the remote repository
➤ YN0001: │ Error: launchdarkly-js-sdk-common@https://github.com/ExplodingCabbage/js-sdk-common.git#commit=eb708e021fe33b38e7c4f5a4b970b7dfe718b24f: ENOENT: no such file or directory, rename '/tmp/xfs-b352d5b1/┌────────────────────────────────────────────┐
│                                            │
│   Destination: ./dist/ldclient-common.js   │
│   Bundle Size:  115.61 KB                  │
│   Minified Size:  37.88 KB                 │
│   Gzipped Size:  14.03 KB                  │
│                                            │
└────────────────────────────────────────────┘
┌──────────────────────────────────────────────┐
│                                              │
│   Destination: dist/ldclient-common.cjs.js   │
│   Bundle Size:  109.34 KB                    │
│   Minified Size:  44.18 KB                   │
│   Gzipped Size:  14.7 KB                     │
│                                              │
└──────────────────────────────────────────────┘
┌─────────────────────────────────────────────┐
│                                             │
│   Destination: dist/ldclient-common.es.js   │
│   Bundle Size:  109.17 KB                   │
│   Minified Size:  44.03 KB                  │
│   Gzipped Size:  14.66 KB                   │
│                                             │
└─────────────────────────────────────────────┘
launchdarkly-js-sdk-common-3.3.3.tgz' -> '/tmp/xfs-b352d5b1/package.tgz'
➤ YN0000: └ Completed in 33s 962ms
➤ YN0000: Failed with errors in 33s 966ms
```

I didn't fancy spending time debugging that and didn't see any guidance on how to do manual testing in CONTRIBUTING.md, so I thought I'd leave it to y'all to manually validate the change, if that's alright with you!

### CODE REVIEW

File: src/__tests__/LDClient-events-test.js
Comment (bwoskow-ld) [inline src/__tests__/LDClient-events-test.js:277]: There's a failing GitHub check because this is missing a trailing semi-colon.

Comment (bwoskow-ld) [reply src/__tests__/LDClient-events-test.js:277]: Actually, rather than adding an assertion to this existing test case, it would be preferable to have a new test case. That way we can trigger multiple events and additionally verify that the Do-Not-Track message is only logged once.

File: src/messages.js
Comment (bwoskow-ld) [inline src/messages.js:114]: Minor: let's rephrase this as:

```suggestion
  return 'LaunchDarkly events will not be sent because this browser's Do Not Track setting is enabled.';
```

File: src/index.js
Comment (bwoskow-ld) [inline src/index.js:127]: I'm not sure that we want to log at the _warning_-level here -- this message is primarily intended for developers as opposed to end-users, and yet it is equally likely that end users will encounter the message. End users shouldn't have the noise of seeing this warning on websites using LaunchDarkly.

My preference would be to lower this to a debug-level log. I expect that this would have been sufficient to save you and your colleague your lost time -- do you agree?

```suggestion
        logger.debug(messages.doNotTrackEnabled());
```

**Review by bwoskow-ld — CHANGES_REQUESTED** [review]
Hi @ExplodingCabbage,

Thanks for the contribution! I left some small feedback, but overall this seems like a reasonable addition.

File: src/messages.js
Comment (ExplodingCabbage) [reply src/messages.js:114]: Fine by me!

File: src/index.js
Comment (ExplodingCabbage) [reply src/index.js:127]: Hmm. No, I don't think this makes sense:

* Debug-level logs are invisible in the Chrome console by default; you have to enable "verbose" mode from a little dropdown at the top of the console to be able to see them
* I think the logger you create in the JS client SDK gets created with a max level of "info" and so if you change this to `.debug` it won't actually log anything at all

I'm not sure the developers vs end users distinction makes much sense to me either. Anyone opening the console and looking at what's been logged can basically be presumed to be a developer, no? Or at least, if they're a real user, they've made a deliberate choice to pop open the dev tools and look under the hood. It's not like the logger is going to put these messages onto the page.

Finally, some other messages that feel similar in character to this one (like "localStorage is unavailable") already get logged at `.warn` level, so using `.warn` here seems consistent with the precedent you've got for when to use which log levels.

File: src/__tests__/LDClient-events-test.js
Comment (ExplodingCabbage) [reply src/__tests__/LDClient-events-test.js]: > That way we can trigger multiple events and additionally verify that the Do-Not-Track message is only logged once.

Makes sense.

> it would be preferable to have a new test case.

Hmm. Thing is, this test case would basically be a copy-and-paste of the existing test case but with extra lines. If one test is going to basically be a subset of the other, it probably makes sense to just have the larger one?

With that in mind I've extended the existing test case to make a second `.identify` call and then check that the logger output is unchanged.

Possibly we've got slightly different philosophies about test suite design and this won't seem like the right solution to you; if so, feel free to tweak as you see fit! *"Allow edits by maintainers"* is switched on. :)

**Comment by ExplodingCabbage** [comment]
@bwoskow-ld  I've addressed all points in one way or another - let me know your thoughts!

File: src/index.js
Comment (eli-darkly) [reply src/index.js:127]: > I think the logger you create in the JS client SDK gets created with a max level of "info" and so if you change this to .debug it won't actually log anything at all

The _default_ logger that you get if you don't customize anything has a level of "info", but you can customize that. The intention is that if a developer is trying to investigate some unexpected behavior and wants to see detailed logging, they would set it to "debug".

This particular message does seem to me like something we would _not_ want to emit by default, because it's not of any possible use to a regular user— that is, they already know that they turned on do-not-track, and it would just be saying that it is behaving as expected. The "localStorage is unavailable" message is different, because there it is saying the SDK functionality is _unable_ to work as desired.

Comment (eli-darkly) [reply src/index.js:127]: I realize that from the _developer's_ point of view, the distinction I'm making there is not so clear, because analytics events are a thing you want to be able to use and they're being blocked. But the effect of that is only visible on the back end— it doesn't have any effect on the user's experience, whereas local storage being disabled potentially does have a noticeable effect (long load times due to no caching). Maybe I am splitting hairs, but that's the thinking that was behind the `warn` level there.

Comment (eli-darkly) [reply src/index.js:127]: As for whether "anyone opening the console and looking at what's been logged can basically be presumed to be a developer"— the concern with appropriateness of `warn` logging isn't that the log would disrupt anything for normal users, but simply that if someone who does know how to use the console gets curious and looks at the log, something that looks like a _warning_ message may give an misleading sense of something being broken. I would be fine with this being either `info` or `debug`, but some browsers (such as Chrome) will show `warn` messages in a style that connotes "potential problem, please pay attention to this" and that's really not applicable in this case. I presume in this particular case you would have noticed an `info` message in the console too.

Comment (ExplodingCabbage) [reply src/index.js:127]: > The _default_ logger that you get if you don't customize anything has a level of "info", but you can customize that. The intention is that if a developer is trying to investigate some unexpected behavior and wants to see detailed logging, they would set it to "debug".

Okay, but that's still requiring additional reading and debugging effort from a dev who's trying to get started. I don't think I even realised that enabling more verbose logging was possible prior to it coming up here, and it would've been (and was) low on my list of ideas to try when observing `.identify` calls not triggering requests. Also, I don't think it's covered outside of the API docs currently, which is a long way to stray for someone who's trying to follow the Getting Started guide or fiddling with the existing LaunchDarkly integration in their company's website for the first time and wondering why it's not working in their dev environment. Perhaps you've got a better gauge of it than I do, but I would be surprised if more than a tiny minority of users of the SDK have ever tweaked the logger config to aid their debugging.

It seems much friendlier to just put the explanation for why stuff doesn't work right there in the console for them, without them having to put in extra effort to solicit debugging information. I'd expect to need to enable debug logging if I wanted to see minutiae like every event and every flag evaluation being logged, but a big slice of the SDK's functionality being disabled feels doesn't feel like a debug-level event to me.

> it's not of any possible use to a regular user

I still don't get this - what do we mean by a regular user here and why is such a person looking in the console in the first place?

> they already know that they turned on do-not-track

You'd think so, but this isn't a safe assumption. Neither I nor my colleague knew; in both cases it was turned on by privacy plugins we'd installed in our browser, rather than due to us specifically enabling it in our browser settings. Also, even if they know it's turned on, they don't necessarily know LaunchDarkly respects it.

> The "localStorage is unavailable" message is different, because there it is saying the SDK functionality is unable to work as desired.

To my mind these are similar; in both cases, the SDK is running in an environment where some of its functionality is disabled, and so in both cases it makes sense to emit a warning so that when we're in a dev environment the developer sees that some functionality is turned off and why.

Comment (ExplodingCabbage) [reply src/index.js:127]: > I presume in this particular case you would have noticed an info message in the console too.

Yes, this would work - but I think, actually, that the default logger level is *warn* rather than *info* as I had previously stated... so I think making this info-level would again hit the problem that it's not logged by default.

Comment (eli-darkly) [reply src/index.js:127]: Sorry, my mistake - the default level in this case is indeed `warn`. It is `info` for most of our SDKs, but it's deliberately higher for the browser for the reasons I stated. However, your original statement that it's not possible to enable debug logging at all was incorrect. I'm going to defer to the rest of the team when it comes to further discussion of what is or is not reasonable to show by default.

About the documentation, though— I agree that the docs should say more about logging, since https://docs.launchdarkly.com/sdk/client-side/javascript doesn't currently have a section for logging as the other SDKs do.

However, in terms of documentation specifically about do-not-track— that same page _does_ have a warning callout that says:

> Do Not Track and ad blocking software
> The JavaScript SDK respects the Do Not Track events header. If an end-user has Do Not Track enabled in their browser, the SDK does not send analytics events for flag evaluations or goals. In addition, ad-blocking software may block analytics events from being sent. This does not impact feature flag evaluations.

You wouldn't see that if you were _only_ looking at the very basic Getting Started section, but the same is true of lots of areas of SDK functionality. Each SDK has a main docs page like this, which highlights some important points and then links to more detailed information on each feature as well as to the API docs.

Comment (bwoskow-ld) [reply src/index.js:127]: Yeah -- all in all let's change this to an `info`-level message. @ExplodingCabbage

File: src/__tests__/LDClient-events-test.js
Comment (bwoskow-ld) [reply src/__tests__/LDClient-events-test.js]: Yeah, this very well could be due to slight differences in test philosophies. The perspective I'm coming from is that, in the case that the test fails, I want the test name to denote what went wrong. That wouldn't be the case here -- we'd be testing the behavior of the logging, rather than that the SDK "does not send an identify event if doNotTrack is set". Also the behavior we're testing isn't specific to identify events -- for example we also have a test in this file to verify that track events aren't sent if Do-Not-Track is enabled.

As for the copy-paste concern, the test case I'm thinking of is _slightly_ different -- effectively we'd have two calls to generate events before making any assertions.. For example, two `client.identify` calls, or maybe one `client.identify` and one `client.track`, followed by just a single assertion for the single logged message.

File: src/index.js
Comment (ExplodingCabbage) [reply src/index.js:127]: Well, I can't stop you, and you're welcome to take over the PR and implement what you like. I'm not going to do it myself because IMO it renders the whole PR more or less pointless.

**Pull request closed without merging**
